### PR TITLE
quarantine: automatically delete quarantined files after specified time

### PR DESCRIPTION
**Pull request by lpranam**
Signed-off-by: Pranam Lashkari <lpranam@collabora.com>
Change-Id: Idabef2aec46d6b7c194ef06f4235b60d6ad29e4a

* Target version: master 

### Checklist

- [ ] Code is properly formatted
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by tml1024** [comment]
Looks fine, but will check in a "make run".

File: wsd/QuarantineUtil.cpp [hunk 1]
```diff
@@ -78,6 +78,48 @@ namespace Quarantine
         return size;
     }
 
+    void deleteExpiredFiles(std::vector<std::string> &files)
+    {
+        if (!isQuarantineEnabled())
+            return;
+
+        // parameter is used in this function to avoid creating and sorting file list multiple times
+        if(files.empty())
```
Comment (mmeeks) [inline wsd/QuarantineUtil.cpp:87]: We have a single call site surely ? and by the time we get here it has a list of all the files surely (?)

File: wsd/QuarantineUtil.cpp [hunk 2]
```diff
@@ -78,6 +78,48 @@ namespace Quarantine
...
+        }
+
+        std::size_t timeLimit = LOOLWSD::getConfigValue<std::size_t>("quarantine_files.expiry_min", 30);
+        const auto timeNow = std::chrono::system_clock::now();
+        const auto ts = std::chrono::duration_cast<std::chrono::seconds>(timeNow.time_since_epoch()).count();
+
+        std::size_t index = 0;
+        while (!files.empty() && index < files.size())
+        {
+            FileUtil::Stat f(LOOLWSD::QuarantinePath + files[index]);
```
Comment (mmeeks) [inline wsd/QuarantineUtil.cpp:100]: In general it's better to do the syscall just once to stat the file to get its information; can we not factor out some 'bool deleteFile (name, timeNow, &totalSize); type function that we can scan do a single pass through the files with ? =)

File: wsd/QuarantineUtil.cpp [hunk 3]
```diff
@@ -90,6 +132,8 @@ namespace Quarantine
 
         std::sort(files.begin(), files.end());
 
+        deleteExpiredFiles(files);
+
         std::size_t currentSize = quarantineSize();
         std::size_t index = 0;
         while (currentSize >= sizeLimit && !files.empty())
```
Comment (mmeeks) [inline wsd/QuarantineUtil.cpp:139]: So doing a single state in this loop below; and having some nice intermediates to flag what we are detecting and doing some easy-to-read logic around that ?

bool isExpired = <timestamp-check>
bool isInUse = hardlink-count > 1

if (isExpired || !inUse || totalSize > whatever) ...
   erase.

**Review by mmeeks — COMMENTED** [review]
A few thoughts; it looks like it would work - so not against getting it in as-is; but can improve maintainability here quite easily I think.

**Review by timar — APPROVED** [review]

**Pull request merged**